outgoing-webhook: Catch jsonable error in fail_with_message.

ligmitz opened the pull request:
Currently if the server timeouts on hitting a url, the fail_with_message
sends a message to the recipients. But if the bot gets deactivated
just after the failure happened, the fail_with_message returns a
JsonableError from send_response_message. This error piles up
till the queue handler. Hence catch the error in fail_with_message.

Fixes #19101.

**Testing plan:** <!-- How have you tested? -->
I am not sure how to reproduce the behaviour in a test to deactivate the bot after the timeout response, hence added no-coverage to the block for now. 

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

ligmitz committed:
outgoing-webhook: Catch jsonable error in fail_with_message.

Currently if the server timeouts on hitting a url, the fail_with_message
sends a message to the recipients. But if the bot gets deactivated
just after the failure happened, the fail_with_message returns a
JsonableError from send_response_message. This error piles up
till the queue handler. Hence catch the error in fail_with_message.

Fixes #19101.

timabbott commented:
For automated testing, you can use `mock.patch` to wrap `fail_with_message` to first call `do_deactive_user` and then call the original function in the scope of a single test.  (For manual testing, you can add a `time.sleep(30)` and then use the time to deactivate the user first)

ligmitz commented:
@timabbott This is ready for a review.

timabbott commented on zerver/tests/test_outgoing_webhook_system.py:
```diff
@@ -162,6 +170,46 @@ def test_fail_request(self) -> None:
+        mock_event = self.mock_event(bot_user)
+        service_handler = GenericOutgoingWebhookService("token", bot_user, "service")
+
+        def side_effect(*args: Any, **kwargs: Any) -> None:
```
We should pick a more descriptive name for this.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -210,7 +210,10 @@ def fail_with_message(event: Dict[str, Any], failure_message: str) -> None:
+    try:
+        send_response_message(bot_id=bot_id, message_info=message_info, response_data=response_data)
+    except JsonableError:
+        pass
```
The original issue report suggested doing this in `send_response_message`; how did you think about that decision?  I'm not sure what's correct.

I do note that `notify_bot_owner` likely is vulnerable to the same bug if the bot owner is deactivated; it looks like `process_success_response` already catches JsonableError (And then calls `send_response_message`).

So maybe we just need this in both `notify_bot_owner` and here; with a test for the other case.

ligmitz committed:
outgoing-webhook: Catch jsonable error in notify_bot_owner.

Currently if the server timeouts on hitting a url, the notify_bot_owner
sends a message to the bot owner. But if the bot onwer gets deactivated
just after the event failure happened, the notify_bot_owner returns a
JsonableError from send_response_message.  Hence catch the error in
notify_bot_owner.

Fixes #19101.

timabbott commented on zerver/tests/test_outgoing_webhook_system.py:
```diff
@@ -162,6 +170,46 @@ def test_fail_request(self) -> None:
+        )
+
+        assert bot_user.bot_owner is not None
+        self.assertEqual(bot_owner_notification.recipient_id, bot_user.bot_owner.recipient_id)
```
These asserts seem to be in the wrong order; we should probably assert that the bot has an owner at the very top.

ligmitz replied:
Makes sense. Will do.

ligmitz replied:
@timabbott Updated the assert statement for the first commit. For the second commit, the current position of the assert statement is necessary otherwise it throws a mypy error:
```
mypy      | zerver/tests/test_outgoing_webhook_system.py:222: error: Argument 1 to "do_deactivate_user" has incompatible type "Optional[UserProfile]"; expected "UserProfile"  [arg-type]
mypy      | 
mypy      | See https://zulip.readthedocs.io/en/latest/testing/mypy.html for debugging tips.
```

timabbott replied:
ok.

timabbott commented on zerver/tests/test_outgoing_webhook_system.py:
```diff
@@ -210,6 +211,43 @@ def deactivate_bot_first(*args: Any) -> None:
+            self.assertEqual(
+                m.output,
+                [
+                    f'WARNING:root:Message http://zulip.testserver/#narrow/stream/999-Verona/topic/Foo/near/ triggered an outgoing webhook, returning status code 408.\n Content of response (in quotes): "{final_response.text}"'
```
Something is wrong with this error message: We shouldn't have a `near:` link without a message ID. Do you understand the bug?

ligmitz replied:
Hmm, I am not sure why us it so. This test is adapted from a similar test just above.

ligmitz replied:
We have `id` as empty string in the message content for `mock_event`.

timabbott commented on zerver/tests/test_outgoing_webhook_system.py:
```diff
@@ -210,6 +211,43 @@ def deactivate_bot_first(*args: Any) -> None:
+        with mock.patch.object(
+            service_handler, "session"
+        ) as session, expected_fail as mock_fail, self.assertLogs(level="WARNING") as m:
+            session.post.return_value = ResponseMock(408)
```
Can you explain what's happening a bit more here?  E.g. why is it a 408 error?

It might be cleaner to just have this be the case where both the bot and the bot owner are deactivated, since that's what is most likely to happen in practice (as deactivating a user cascades to deactivating all their bots).

ligmitz replied:
408 is timeout error.

ligmitz commented:
Hmm, this is strange. Ideally, the test should not cause a failure message to be sent to the stream since the bot onwer and hence the bot is deactivated which works as expected in manual testing but not in automated testing. But this also does not triggers a no coverage error for the case in the automated test. I am not sure why is it so.

ligmitz committed:
outgoing-webhook: Rename mock_events in outgoing-webhook tests.

Rename `mock_event` to `mock_stream_message_event`.

ligmitz commented:
@timabbott I have updated the pull request.
Changes:
* Renamed `mock_event` to `mock_stream_message_event`.
* Added test for bot the events type.

The issue says when the url hits the timeout, we want the error handling so I have mocked the 408 timeout status code in the tests.
